### PR TITLE
BUGFIX: Trying to run any command defined in `amixer` failes.

### PR DESCRIPTION
**Pull request by equwal**
BUGFIX: Trying to run any command defined in `amixer` failes.
# Reproduction
  Run `(amixer:amixer-Front-toggle)`, which should toggle a front speaker.

# Result

```common-lisp
; The value
;   NIL
; is not of type
;   VECTOR
;    [Condition of type TYPE-ERROR]
;
; Backtrace:
    (AMIXER::VOLCONTROL NIL "Front" "toggle")
    (AMIXER:AMIXER-FRONT-TOGGLE NIL)
```

To fix this bug, amixer has been moved to using the `-c` argument instead of
`-D`. StumpWM also has a breaking change to `run-shell-command`.

### CODE REVIEW

**Comment by dmb2** [comment]
Can you confirm this is a bug if you call the internal symbol (amixer::amixer-Front-toggle)?

The only bugfix worthy change I see is the switch in command line switches.

**Comment by equwal** [comment]
> Can you confirm this is a bug if you call the internal symbol (amixer::amixer-Front-toggle)?

Yes, I have just tested this. I also just reinstalled my operating system, are you able to reproduce?

> The only bugfix worthy change I see is the switch in command line switches.

Fair enough. That is now the only commit in this pull.

File: media/amixer/package.lisp [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 (defpackage #:amixer
   (:use #:cl :stumpwm)
+  (:import-from :alexandria :with-gensyms)
```
Comment (kriyative) [inline media/amixer/package.lisp:5]: This import isn't being used.

File: media/amixer/amixer.lisp [hunk 2]
```diff
@@ -9,12 +9,12 @@
...
 ;;; Code:
-(defvar *default-device* "default")
+(defvar *default-device* 0)
 
 (defun volcontrol (device channel amount)
   (let* ((output (run-shell-command
-                  (concat "amixer -D "
-                          (or device *default-device*)
+                  (concat "amixer -c "
+                          (write-to-string (or device *default-device*))
```
Comment (kriyative) [inline media/amixer/amixer.lisp:17]: I find the symbolic device name to be more readable/useful. Is there any issue continuing to use the `-D` arg?

**Comment by kriyative** [comment]
I use the `amixer` module regularly on Ubuntu Linux 18.04.4 LTS and I've not had any issue using the `-D ..` form of the command as the module is currently implemented.

**Comment by kriyative** [comment]
One thing I would add is that in order to have `amixer` (the Linux binary) correctly manage the audio controls, you may have to bind `amixer::*default-device*` in Lisp to the appropriate symbolic name. Keep in mind the control name "Front" may or may not exist on every audio device. In my particular system, the PulseAudio managed device is called "pulse" and it has controls named "Master" and "Capture" (for the mic).

**Comment by m-g-r** [comment]
> BUGFIX: Trying to run any command defined in `amixer` failes.
> # Reproduction
> 
> Run `(amixer:amixer-Front-toggle)`, which should toggle a front speaker.
> # Result
> 
> ```lisp
> ; The value
> ;   NIL
> ; is not of type
> ;   VECTOR
> ;    [Condition of type TYPE-ERROR]
> ;
> ; Backtrace:
>     (AMIXER::VOLCONTROL NIL "Front" "toggle")
>     (AMIXER:AMIXER-FRONT-TOGGLE NIL)
> ```
> 
> To fix this bug, amixer has been moved to using the `-c` argument instead of
> `-D`. StumpWM also has a breaking change to `run-shell-command`.

### This fix is in correct, as the error happens later and is not connected to the `-c` or `-D` options.

If you debug `(amixer:amixer-Front-toggle)`, you'll see that the call to `amixer` translates correctly:
```
Evaluating call:
  (RUN-SHELL-COMMAND
   (CONCAT "amixer -D " (OR DEVICE *DEFAULT-DEVICE*) " sset " CHANNEL " "
           (OR AMOUNT "toggle"))
   T)
With arguments:
  "amixer -D default sset Front toggle"
  T
```
so that is okay. The type error `The value NIL is not of type VECTOR` happens later in the AREF in `(string= (aref percent-status 1) "off")`!

(Side note: Be not misled that a simple `(aref nil 1)` gives an error saying ARRAY instead VECTOR; if you let SBCL compiler have a go it first by wrapping it in a LAMBDA `((lambda () (aref nil 1)))` it transforms it to a single dimensional AREF and complains about VECTOR instead of ARRAY.)

### So why did this change work for you (and other PulseAudio users, while still others were happy as it was)?

So the real problem is that the output of `amixer` doesn't contain the expected on/off message that the call to `cl-ppcre:scan-to-strings` expected. And this is most probably because you have a system using PulseAudio where the default device that you get with `amixer -D default` or just `amixer` withouth any `-d` and `-c` has not as much functionality than a direct audio device.  You get a device like this:
```
mgr@pulseaudio:~ $ amixer -D default info
Card default 'pulse'/'PulseAudio'
  Mixer name	: 'PulseAudio'
  Components	: ''
  Controls      : 4
  Simple ctrls  : 2
```
while `-c 0` gives you directly a hw audio device like this:
```
mgr@pulseaudio:~$ amixer -c 0 info
Card hw:0 'PCH'/'HDA Intel PCH at 0xf1240000 irq 133'
  Mixer name	: 'Conexant CX20753/4'
  Components	: 'HDA:14f15111,17aa2238,00100101 HDA:80862809,80860101,00100000'
  Controls      : 56
  Simple ctrls  : 15
```
But if you have direct ALSA without PulseAudio both calls give you the same direct hw device:
```
mgr@alsa:~$ amixer -D default info
Card default 'PCH'/'HDA Intel PCH at 0xfb500000 irq 52'
  Mixer name	: 'Realtek ALC892'
  Components	: 'HDA:10ec0892,10438469,00100302'
  Controls      : 51
  Simple ctrls  : 22
mgr@alsa:~$ amixer -c 0 info
Card hw:0 'PCH'/'HDA Intel PCH at 0xfb500000 irq 52'
  Mixer name	: 'Realtek ALC892'
  Components	: 'HDA:10ec0892,10438469,00100302'
  Controls      : 51
  Simple ctrls  : 22
```

Toggling itself works fine with PulseAudio's default device works fine, even untoggling doesn't leave outputs muted (see the discussion in https://github.com/stumpwm/stumpwm-contrib/pull/184) but changing the audio volume fails:
```
mgr@pulseaudio:~$ amixer -D default sset Master playback 2db+
amixer: Invalid command!
```
So this explains why changing from `-D default` to `-c 0` worked for you: You've controlled a different device. Took a bit of time to figure this out for me as well.

### TLDR; / proposal

This is a problem of PulseAudio but making programs aware of the idiosyncrasies of PulseAudio usually let's them fall in the same complexity trap of PulseAudio, and will probably not be good for users of just bare ALSA without PulseAudio.

Still, you can make this nicer by just translating the bare device numbers of `-c` to device names for `-D` by prefixing them with `hw:`, so that `-c 0` becomes `-D hw:0`.

So, use `-D` but implement this handling:
1. if `*default-device*` is a number, translate it to string by prefixing with `hw:`
2. if it is a string, leave it as it is
3. if it is NIL, replace by "default" (or leave out the `-D` in the call altogether)
(error or default case for other values.)

This should work for all and make both, @equwal and @kriyative, happy. I'll change to this behavior in the minimialistic stump-volume-control as well. 

I can make a commit for just this but the maintainer for the amixer module should make the parsing of `amixer` more robust.

**Comment by m-g-r** [comment]
@equwal I cannot really test your case as my sound device does not have any "Front" channel (or control), but "Master" and "Speaker" work for me with PR #194:
```
mgr@pulseaudio:~$ amixer -D hw:0 sset Master toggle
Simple mixer control 'Master',0
  Capabilities: pvolume pvolume-joined pswitch pswitch-joined
  Playback channels: Mono
  Limits: Playback 0 - 74
  Mono: Playback 48 [65%] [-26.00dB] [on]
mgr@pulseaudio:~$ amixer -D hw:0 sset Speaker toggle
Simple mixer control 'Speaker',0
  Capabilities: pvolume pswitch
  Playback channels: Front Left - Front Right
  Limits: Playback 0 - 74
  Mono:
  Front Left: Playback 74 [100%] [0.00dB] [off]
  Front Right: Playback 74 [100%] [0.00dB] [off]
```
Can you share an output of `amixer -D hw:0 sset Front toggle` on your system? For me, as I do not have any "Front" the output is like this and thus has no on/off info:
```
mgr@pulseaudio:~$ amixer -D hw:0 sset Front toggle
amixer: Unable to find simple control 'Front',0

mgr@pulseaudio:~$ echo $?
1
```
And for "PCM" which has no mute on my system the output is just completely empty and doesn't even return an error code:
```
mgr@pulseaudio:~$ amixer -D hw:0 sset PCM toggle
mgr@pulseaudio:~$ echo $?
0
```

So the PR #194 makes it nicer but doesn't fix the error behavior of the amixer module and might not fix the problem of @equwal. But I cannot debug this further and give this work back to others. I have got enough of this already. :)

**Comment by equwal** [comment]
I'm going to close this due to the request by m-g-r which keeps support for pulseaudio and amixer.

https://github.com/stumpwm/stumpwm-contrib/pull/194

**Comment by equwal** [comment]
After having completely deleted pulseaudio from my system I can confirm that this works fine without pulseaudio, and is thus a bug in the support for pulseaudio.

**Comment by m-g-r** [comment]
> After having completely deleted pulseaudio from my system I can confirm that this works fine without pulseaudio, and is thus a bug in the support for pulseaudio.

Haha, the same solution I took when installing a new system. :-)
